### PR TITLE
Core/Script/Spell: Fix bug when Mirror Images could not start attack standin…

### PR DESCRIPTION
**Pull request by Ofinka**
fixes problem with attack to standing characters #14877

### CODE REVIEW

**Comment by Keader** [comment]
Hello Ofinka \o
First thanks for the fix. Just test it, it's working. However're a little problem. The spell says the mirrors must attack the target of greatest threat, with its fix this is not happening. If you use a mirror image of the dummy ironforge (.tele themilitaryward), will notice that the images will attack random targets

**Comment by Goatform** [comment]
I think that images should to attack same target if they detect him, and if owner start attacking another target, they should to switch target and attack owner's target.

**Comment by tkrokli** [comment]
@Keader : if you read the code, it is only written to do 1 of 2 things; either attack owner's victim (`owner->GetVictim()`) or attack a random victim (`AnyUnfriendlyUnitInObjectRangeCheck`). If you want the mirror images to attack a target/victim with highest threat, some extra code needs to be added.
(I don't know if it can be added in the same place, though.)

**Comment by Ofinka** [comment]
ye guys i realize victim choosing of this code is based on gargoyle targeting system.. i dont know why did I think it is right to do that this way. I will try to update this PR asap with threat victim finding if I find some way how to do that effectively

**Comment by Keader** [comment]
The real problem is that sometimes you own target, use the spell and 1 or 2 images  attack other random target. If we ever attack the mage  target  maybe the prejudice "not to attack the first in the list of threat" goes unnoticed.

**Comment by Killyana** [comment]
Some informations from wowwiki:
- The Mirror Images inherit the Mage's threat lists
- Patch 3.0.8 (2009-01-20): In addition, Mirror Images will no longer have excessive threat values assigned to them when they are created.
-  Patch 3.1.2 (2009-05-19): The images summoned by this spell will now target the creature that most hates the Mage, and should no longer cast Fire Blast or Frostbolt on targets that are affected by crowd control debuffs that break immediately on damage unless they are already casting these spells when crowd control is applied.
- The mirror images will no longer complete casts of Frostbolt on targets which are Polymorphed at the time their Frostbolt channel finishes.

Other informations from forums:
- They attack same target as you, and follow you always - the only way they behave, and only way to control them
- Didn't read everything but, they seem to attack whatever you have most threat on. This is why if you pop them before a boss is pulled, they attack nothing as you have threat on nothing till you first attack the boss yourself. In my experience they change target when you Invi to drop threat and such.

**Comment by Goatform** [comment]
Using AnyUnfriendlyUnitInObjectRangeCheck  isn't good solution for this problem because they shouldn't attack random targets. They just should to attack your target once you start attacking it.

**Comment by Ofinka** [comment]
try this update pls, any feedback is welcomed

**Comment by Keader** [comment]
@jackpoz Do you have any suggestions for how to let code less?

**Comment by Keader** [comment]
@Ofinka The CC mechanics are without work. Whenever you are in CC they attack, totally ignoring the mechanics.

**Comment by Ofinka** [comment]
works for me

**Comment by Keader** [comment]
Work for e too :+1:

**Comment by mik1893** [comment]
does Init()  need to be called also in the reset? (asking)
Also, please squash the commits into one

File: src/server/scripts/Pet/pet_mage.cpp [hunk 1]
```diff
@@ -40,20 +50,185 @@ class npc_pet_mage_mirror_image : public CreatureScript
         {
             npc_pet_mage_mirror_imageAI(Creature* creature) : CasterAI(creature) { }
 
+            void Init()
+            {
+                Unit* owner = me->GetCharmerOrOwner();
+
+                std::list<Unit*> targets;
+                Trinity::AnyUnfriendlyUnitInObjectRangeCheck u_check(me, me, 30.0f);
+                Trinity::UnitListSearcher<Trinity::AnyUnfriendlyUnitInObjectRangeCheck> searcher(me, targets, u_check);
```
Comment (mik1893) [inline src/server/scripts/Pet/pet_mage.cpp]: if(!targets.empty())   
{

Comment (ccrs) [inline src/server/scripts/Pet/pet_mage.cpp]: not necessary

**Comment by Ofinka** [comment]
@mik1893  I dont see calling Init() fn in reset()

**Comment by jackpoz** [comment]
is this complete, tested and ready to be merged ? 
I see some required codestyle changes but I can take care of them when merging it.

**Comment by Ofinka** [comment]
@jackpoz  Ye, I think so..

**Comment by ShinDarth** [comment]
any news?

**Comment by ShinDarth** [comment]
any news? @jackpoz ?

**Comment by ShinDarth** [comment]
I've been testing this for days on https://github.com/TrinityCore/TrinityCore/commit/cc685ac334335d9f6a25a7c1c739a905f919328f  with players and they are ginving me only positive feedback about it. I think it may be merged.

File: src/server/scripts/Pet/pet_mage.cpp [hunk 2]
```diff
@@ -40,20 +50,185 @@ class npc_pet_mage_mirror_image : public CreatureScript
...
+                            // If this player is closer than the previous one, update it
+                            if (me->GetDistance((*iter)->GetPosition()) < me->GetDistance(nearestPlayer->GetPosition()))
+                                nearestPlayer = (*iter);
+                        }
+                    }
+                }
+                // Prioritize units with threat referenced to owner
+                if (highestThreat > 0.0f && highestThreatUnit)
+                        me->Attack(highestThreatUnit, false);
+
```
Comment (jackpoz) [inline src/server/scripts/Pet/pet_mage.cpp]: this empty line is not needed

File: src/server/scripts/Pet/pet_mage.cpp [hunk 3]
```diff
@@ -40,20 +50,185 @@ class npc_pet_mage_mirror_image : public CreatureScript
...
+                }
+
+                if (me->HasUnitState(UNIT_STATE_CASTING))
+                    return;
+
+                if (!owner)
+                    return;
+
+                // assign target if image doesnt have any or the target is not actual
+                if (me->GetVictim() && me->GetVictim() != target || !me->GetVictim())
```
Comment (jackpoz) [inline src/server/scripts/Pet/pet_mage.cpp]: this looks like the same as
```
if (!target || me->GetVictim() != target)
```

**Comment by jackpoz** [comment]
thank you for the fix

**Pull request merged**